app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/hplus/HPlusHandlerThread.java

```diff
@@ -152,30 +160,56 @@ public void quit() {
     }
 
     public void sync() {
+        LOG.info("HPlus: Starting data synchronization");
+
         mGetSleepTime.setTimeInMillis(0);
         mGetDaySlotsTime.setTimeInMillis(0);
         mGetDaySummaryTime.setTimeInMillis(0);
         mLastSleepDayReceived.setTimeInMillis(0);
+        mHelloTime = GregorianCalendar.getInstance();
+        mHelloTime.add(Calendar.SECOND, HELLO_PERIOD);
 
         mSlotsInitialSync = true;
         mLastSlotReceived = -1;
         mLastSlotRequested = 0;
         mCurrentDaySlot = null;
         mDaySlotRecords.clear();
 
-        TransactionBuilder builder = new TransactionBuilder("startSyncDayStats");
+        try {
+            if(!mHPlusSupport.isConnected())
+                mHPlusSupport.connect();
 
-        builder.write(mHPlusSupport.ctrlCharacteristic, new byte[]{HPlusConstants.CMD_GET_DEVICE_ID});
-        builder.write(mHPlusSupport.ctrlCharacteristic, new byte[]{HPlusConstants.CMD_GET_VERSION});
-        builder.write(mHPlusSupport.ctrlCharacteristic, new byte[]{HPlusConstants.CMD_GET_CURR_DATA});
+            TransactionBuilder builder = new TransactionBuilder("startSyncDayStats");
```

Comment:
I think this pattern

> if (!isConnected)
>   connect()
> new TransactionBuilder()...

should be avoided. Rather get a new `TransactionBuilder` like this:
`TransactionBuilder builder = mHPlusSupport.performInitialized();`
This will make sure that your code using the builder will get executed as soon as the connection is established and all initialization is done.

If anything fails during connection or initialization, your code using the builder will *not* be executed (because you  won't be able to communicate with the device anyway.